File: c++/pointers/r_and_l_value.cpp
/*
 * Copyright (C) 2019-2026 Katarzyna Krasińska
 * Languagium - https://github.com/katheroine/languagium
 * Licensed under GPL-3.0 - see LICENSE.md
 */
// SPDX-License-Identifier: GPL-3.0-or-later

#include <iostream>

int main() {
  int number, value, *p_integer, *int_ptr;

  number = 15; // number is l-value
  p_integer = &number; // p_integer is l-value

  std::cout << "number: " << number << ", *p_integer: " << *p_integer << std::endl;
  std::cout << "value: " << value << ", *int_ptr: " << *int_ptr << std::endl << std::endl;

  value = number; // number is r-value

  std::cout << "number: " << number << ", *p_integer: " << *p_integer << std::endl;
  std::cout << "value: " << value << ", *int_ptr: " << *int_ptr << std::endl << std::endl;

  int_ptr = p_integer; // p_integer is r-value

  std::cout << "number: " << number << ", *p_integer: " << *p_integer << std::endl;
  std::cout << "value: " << value << ", *int_ptr: " << *int_ptr << std::endl << std::endl;
}

File: pascal/pointers/r_and_l_value.pas
program PointersAsRAndLValue;

var
  number, value: integer;
  p_integer, int_ptr: ^integer;

begin
  number := 15; // number is l-value
  p_integer := @number; // p_integer is l-value

  writeln('number: ', number, ', p_integer^: ', p_integer^);

  value := number; // number is r-value

  writeln('value: ', value);

  int_ptr := p_integer; // p_integer is r-value

  writeln('int_ptr^: ', int_ptr^);
end.
